src/Raven.Server/Documents/DocumentTombstoneCleaner.cs

```diff
@@ -33,23 +35,35 @@ public void Initialize()
 
         public void Subscribe(IDocumentTombstoneAware subscription)
         {
-            lock (_locker)
+            _locker.Wait();
+
+            try
             {
                 _subscriptions.Add(subscription);
             }
+            finally
+            {
+                _locker.Release();
+            }
         }
 
         public void Unsubscribe(IDocumentTombstoneAware subscription)
         {
-            lock (_locker)
+            _locker.Wait();
+
+            try
             {
                 _subscriptions.Remove(subscription);
             }
+            finally
+            {
+                _locker.Release();
+            }
         }
 
-        internal bool ExecuteCleanup()
+        internal async Task<bool> ExecuteCleanup()
         {
-            if (Monitor.TryEnter(_locker) == false)
+            if (await _locker.WaitAsync(0) == false)
```

Comment:
Yes. I just focused on moving things to the tx merger instead of checking why do we actually need that lock.